src/GNDStk/gnds-tree.hpp

```diff
@@ -0,0 +1,644 @@
+
+}; // class Tree
+
+// Tree::static_*
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+std::string Tree<MCON,CCON>::static_str1 = "";
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+std::string Tree<MCON,CCON>::static_str2 = "";
+
+
+
+// -----------------------------------------------------------------------------
+// helpers
+// -----------------------------------------------------------------------------
+
+namespace detail {
+
+// A couple of these are templated because we'd prefer not to say inline,
+// and this is a header-only library. (Not that "inline" really means it.)
+
+// warning_tree_io_name
+template<class unused>
+void warning_tree_io_name(
+   const std::string &op,
+   const std::string &file,
+   const std::string &format,
+   const std::string &name
+) {
+   warning(
+      "Tree " + op + " called with "
+      "format::" + format + " and "
+      "filename \"" + file + "\",\n"
+      "but the filename extension does not appear to be one for " +
+       name + "."
+   );
+}
+
+// warning_tree_io_data
+template<class unused>
+void warning_tree_io_data(
+   const format f,
+   const std::string &ch,
+   const std::string &name
+) {
+   const std::string form =
+      f == format::xml  ? "xml"
+    : f == format::json ? "json"
+    : f == format::hdf5 ? "hdf5"
+    : "?";
+
+   warning(
+      "Tree read called with "
+      "format::" + form + ",\n"
+      "but the file begins with " + ch + ", which suggests " +
+       name + "."
+   );
+}
+
+// format_tree_read
+inline const std::string format_tree_read =
+   "Format format::tree is not allowed in tree.read(). "
+   "Our \"tree\" format\nis intended only for debug writing; "
+   "it is not a first-class file format."
+;
+
+} // namespace detail
+
+
+
+// -----------------------------------------------------------------------------
+// read
+// -----------------------------------------------------------------------------
+
+// read(char *, format)
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool Tree<MCON,CCON>::read(
+   const char * const file,
+   const format form
+) {
+   // ------------------------
+   // Clear current contents
+   // ------------------------
+
+   // Note that this happens even if something below fails. This is reasonable
+   // behavior; an empty tree is a reminder that the read attempt failed.
+   clear();
+
+   // ------------------------
+   // format::tree
+   // Not allowed in read
+   // ------------------------
+
+   // Error; this format isn't allowed for read() (only for write())
+   if (form == format::tree) {
+      error(detail::format_tree_read);
+      return false;
+   }
+
+   // ------------------------
+   // format::xml,json,hdf5
+   // Check: consistent name?
+   // ------------------------
+
+   if (form == format::xml  && has_extension(file) && !endsin_xml (file))
```

Comment:
No worries; if you look closely, you'll see that I don't actually use the extension for that purpose, but peek() for the file magic number instead. (For reading, that is. There's no file magic number if we're *writing* a new file!) File extensions, when available, are useful in terms of emitting warnings about things that look weird but might be intentional. There are lots of things going on in the code you're looking at, and I claim that everything I did was absolutely intentional, and is there for a reason. I can explain in detail. We have read and write; stream input and output; file name in some places but just a stream in others; an optional user-specified format (via an enum class) to allow someone to force the issue, etc. I have logic to deal with what's available in various cases, and do what I believe is the most reasonable thing in all situations. Everything I'm doing is there for a good reason (unless I made a mistake), but one has to look really carefully to see the what/why/how of what happens in every case. Your ".gnds" extension wouldn't cause any problems whatsoever; the system will read the file magic number, and not complain about the extension. (If the magic number was '<' and the file extension was .json, then it would point out the inconsistency. But it doesn't know about .gnds, so won't complain about anything.)